docs/spec_draft.md

```diff
@@ -102,26 +102,191 @@ never produced as outputs.
+More formally, a **StableHLO process** is a combination of:
+1) a StableHLO program, 2) operation statuses (not executed yet,
+already executed), and 3) intermediate values that the process is working on.
+The process starts with input values to the `main` function, progresses through
+the graph of ops updating operation statuses and intermediate values and
+finishes with output values. Further formalization is TBD.
+
+### Parallel execution
+
+StableHLO programs can be executed in parallel, organized into a 2D grid of
+`num_replicas` by `num_partitions` which both have type `ui32`.
+
+In the **StableHLO grid**, `num_replicas * num_partitions` of StableHLO
+processes are executing at the same time. Each process has a unique
+`process_id = (replica_id, partition_id)`, where
+`replica_id ∊ replica_ids = [0, ..., num_replicas-1]` and
+`partition_id ∊ partition_ids = [0, ..., num_partitions-1]` which both have
+type `ui32`. The size of the grid is known statically for every program, and
+the position within the grid is known statically for every process.
+
+Within the grid, the programs can all be the same (in the "Single Program,
+Multiple Data" style), can all be different (in the "Multiple Program, Multiple
+Data" style) or something in between. Each program has access to its position
+within the grid via the `replica_id` and `partition_id` ops.
+
+Within the grid, the processes are mostly independent from each other - they
+have separate operation statuses, separate input/intermediate/output values and
+most of the ops are executed separately between processes, with the exception of
+a small number of collective ops described below.
+
+Given that execution of most of the ops is only using values from the same
+process, it is usually unambiguous to refer to these values by their names.
+However, when describing semantics of collective ops, that is insufficient, and
+we use the notation `name@process_id` to refer to the value `name` within a
+particular process. (From what perspective, unqualified `name` can be viewed as
+a shorthand for `name@(replica_id(), partition_id())`).
+
+The execution order across processes is implementation-defined, except for the
+synchronization introduced by collective ops as described below.
+
+### Collective ops
+
+There are five collective ops in StableHLO: `all_gather`, `all_reduce`,
+`all_to_all`, `collective_permute` and `reduce_scatter`. All these ops split
+the processes in the StableHLO grid into **StableHLO process groups** and
+execute a joint computation within each process group, independently from other
+process groups.
+
+For all processes within each process group, collective ops introduce a
+synchronization barrier, i.e. the execution of a collective op in all these
+processes doesn't start until the execution of all these processes reaches
+the op. Further formalization, e.g. elaborating on what it means for execution
+to reach an op and what happens to a process group if it doesn't happen, is TBD.
+
```

Comment:
Why is this synchronizing the start of the collective operation and not at its end? Why define execution semantics? I'd leave it to something simple like
"
Collective operations constitute potentially blocking synchronization points among processes within a process group.
"

Reply:
Eventually, I'd still like to formalize the semantics to the degree that the producers and consumers find reasonable. The current state of HLO semantics is underdocumented, which makes it hard to reason about it, and I think it's worth investing in more clarity.

This doesn't mean that we are planning to force standardization where implementation-defined behavior makes sense. Overall, the idea is to start with a baseline spec for HLO behavior and then incrementally refine that together as a community.

Given that, I agree about leaving off any kind of details of synchronization for now. I've updated the text along the lines of what you proposed.